src/components/Time/Replay.tsx

```diff
@@ -0,0 +1,309 @@
+import ToolbarDropdown from 'components/ToolbarDropdown/ToolbarDropdown';
+import { UserSettingsActions } from 'actions/UserSettingsActions';
+import { KialiAppAction } from 'actions/KialiAppAction';
+import Slider from 'components/IstioWizards/Slider/Slider';
+import { KialiIcon } from 'config/KialiIcon';
+import { style } from 'typestyle';
+import { toString } from './LocalTime';
+import { DurationDropdownContainer } from 'components/DurationDropdown/DurationDropdown';
+import DateTimeInput from './DateTimeInput';
+import { serverConfig } from 'config';
+
+type ReduxProps = {
+  replayQueryTime: TimeInSeconds;
+  replayWindow: ReplayWindow;
+
+  setReplayQueryTime: (replayQueryTime: TimeInSeconds) => void;
+  setReplayWindow: (replayWindow: ReplayWindow) => void;
+  toggleReplayActive: () => void;
+};
+
+type ReplayProps = ReduxProps & {
+  id: string;
+};
+
+type ReplayState = {
+  isReplaying: boolean;
+  refresherRef?: number;
+  replayFrame: number;
+  replayFrameCount: number;
+  replaySpeed: number;
+};
+
+// key represents replay interval in seconds
+const replayIntervals = {
+  60: '1 minute',
+  300: '5 minutes',
+  600: '10 minutes',
+  1800: '30 minutes'
+};
+
+// key represents refresh interval in seconds
+const replaySpeeds = {
+  1: 'Very Fast',
+  3: 'Fast',
+  5: 'Medium',
+  10: 'Slow'
+};
+
+const defaultReplayInterval = 300; // 5 minutes
+const defaultReplaySpeed = 3; // fast
+const frameInterval = 10; // number of seconds clock advances per frame
+
+const replayStyle = style({
+  display: 'flex',
+  width: '90%',
+  marginTop: '-5px'
+});
+
+const pauseStyle = style({
+  margin: '0 0 0 15px',
+  height: '37px',
+  width: '10px'
+});
+
+const stopStyle = style({
+  margin: '0 10px 0 0',
+  height: '37px',
+  width: '10px'
+});
+
+const sliderStyle = style({
+  width: '100%',
+  margin: '0 20px 0 15px'
+});
+
+export class Replay extends React.PureComponent<ReplayProps, ReplayState> {
+  static getFrameCount = (replayInterval: IntervalInSeconds): number => {
+    return replayInterval > 0 ? Math.floor(replayInterval / frameInterval) : 0;
+  };
+
+  static queryTimeToFrame = (props: ReduxProps): number => {
+    const elapsedTime: IntervalInSeconds = props.replayQueryTime - props.replayWindow.startTime;
+    const frame: number = Replay.getFrameCount(elapsedTime);
+    return frame;
+  };
+
+  static frameToQueryTime = (frame: number, props: ReduxProps): TimeInSeconds => {
+    const queryTime: TimeInSeconds = props.replayWindow.startTime + frame * frameInterval;
+    return queryTime;
+  };
+
+  constructor(props: ReplayProps) {
+    super(props);
+    this.state = {
+      isReplaying: false,
+      refresherRef: undefined,
+      replayFrame: Replay.queryTimeToFrame(props),
+      replayFrameCount: Replay.getFrameCount(props.replayWindow.interval),
+      replaySpeed: defaultReplaySpeed
+    };
+  }
+
+  componentDidMount() {
+    if (!!!this.props.replayWindow.startTime) {
```

Comment:
Wait what.

Reply:
I know, `!!!` is basically equivalent to `!` but I find it more readable to show it's explicitly a negation  of truthy/falsey for non-boolean types.